contracts/fuses/enso/EnsoExecutor.sol

```diff
@@ -0,0 +1,226 @@
+
+    /// @notice Constructs the EnsoExecutor contract
+    /// @param delegateEnsoShortcuts_ Address of the DelegateEnsoShortcuts contract
+    /// @param wethAddress_ Address of the WETH token
+    /// @param plasmaVault_ Address of the PlasmaVault contract
+    /// @dev Reverts if any address parameter is the zero address
+    constructor(address delegateEnsoShortcuts_, address wethAddress_, address plasmaVault_) {
+        if (wethAddress_ == address(0)) {
+            revert EnsoExecutorInvalidWethAddress();
+        }
+        if (delegateEnsoShortcuts_ == address(0)) {
+            revert EnsoExecutorInvalidDelegateAddress();
+        }
+        if (plasmaVault_ == address(0)) {
+            revert EnsoExecutorInvalidPlasmaVaultAddress();
+        }
+        DELEGATE_ENSO_SHORTCUTS = delegateEnsoShortcuts_;
+        WETH_ADDRESS = wethAddress_;
+        PLASMA_VAULT = plasmaVault_;
+    }
+
+    /// @notice Modifier to restrict function access to only the PlasmaVault
+    /// @dev Reverts if msg.sender is not the PlasmaVault address
+    modifier onlyPlasmaVault() {
+        if (msg.sender != PLASMA_VAULT) {
+            revert EnsoExecutorUnauthorizedCaller();
+        }
+        _;
+    }
+
+    /// @notice Executes an Enso shortcut via delegatecall
+    /// @param data_ EnsoExecutorData containing all necessary information for the shortcut execution
+    /// @dev This function:
+    ///      - Delegatecalls into DelegateEnsoShortcuts.executeShortcut()
+    ///      - Transfers all specified tokens back to the sender
+    /// @dev Only callable by the PlasmaVault contract
+    function execute(EnsoExecutorData memory data_) external payable onlyPlasmaVault {
+        if (_balance.assetAddress != address(0)) {
+            revert EnsoExecutorBalanceAlreadySet();
+        }
+
+        // Delegatecall to DelegateEnsoShortcuts.executeShortcut
+        bytes memory delegateCallData = abi.encodeWithSignature(
+            "executeShortcut(bytes32,bytes32,bytes32[],bytes[])",
+            data_.accountId,
+            data_.requestId,
+            data_.commands,
+            data_.state
+        );
+
+        if (data_.wEthAmount > 0) {
+            IWETH9(WETH_ADDRESS).withdraw(data_.wEthAmount);
+        }
+
+        Address.functionDelegateCall(DELEGATE_ENSO_SHORTCUTS, delegateCallData);
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            IWETH9(WETH_ADDRESS).deposit{value: ethBalance}();
+            IERC20(WETH_ADDRESS).safeTransfer(msg.sender, ethBalance);
+        }
+
+        uint256 tokensOutBalance = IERC20(data_.tokensOut).balanceOf(address(this));
+        _balance.assetAddress = data_.tokensOut;
+
+        if (tokensOutBalance > 0) {
+            IERC20(data_.tokensOut).safeTransfer(msg.sender, tokensOutBalance);
+            if (data_.amountOut > tokensOutBalance) {
+                _balance.assetBalance = (data_.amountOut - tokensOutBalance).toUint96();
```

Comment:
point out that this value can be higher than zero but executor doesn't have this tokens now because of cross-chain solutions or other reasons